pepper-apis/src/main/java/org/broadinstitute/ddp/db/ActivityInstanceDao.java

```diff
@@ -379,6 +356,51 @@ public Long getActivityInstanceIdByGuid(Handle handle, String guid) {
         );
     }
 
+    /**
+     * Iterate through list of activity summaries and count up how many questions are answered.
+     *
+     * @param handle            the database handle
+     * @param userGuid          the user guid
+     * @param studyGuid         the study guid
+     * @param activitySummaries the list of activity summaries
+     */
+    public void countActivitySummaryQuestionsAndAnswers(Handle handle, String userGuid, String studyGuid,
+                                                        List<ActivityInstanceSummary> activitySummaries) {
+        ActivityDefStore activityDefStore = ActivityDefStore.getInstance();
+
+        Set<String> instanceGuids = activitySummaries.stream()
+                .map(ActivityInstanceSummary::getActivityInstanceGuid)
+                .collect(Collectors.toSet());
+        Map<String, FormResponse> instanceResponses = handle
```

Comment:
Before, each call to `countQuestionsAndAnswers()` will query for the instance answer data. Key insight here is to do a single query for all the instance answer data we need for the activity summaries rather than individual queries. And now that we have the data loaded, there's an opportunity to leverage those. That will be in upcoming optimizations.

Reply:
Upon closer look locally, the performance here of querying all instances and then doing the counting seems reasonable. The counting part dominants the runtime of the list-activities API, but locally it's averaging < 100 ms, so I think I will leave it for now and do other optimizations instead.